hazelcast/src/main/java/com/hazelcast/map/impl/operation/AddIndexOperation.java

```diff
@@ -93,6 +130,16 @@ public void runInternal() {
 
         recordStore.forEach((dataKey, record) -> {
             Object value = Records.getValueOrCachedValue(record, serializationService);
+            if (recreateMetadata) {
```

Comment:
a simple performance test can be helpful.

Reply:
I don't have any numbers yet, but the additional logic should not cause any performance degradation on maps that do not contain HazelcastJsonValue keys or values. Note that, for them, the size of the metadata store will be 0. So, `recreateMetadata` flag will be false, and the only extra cost we will have will be this if check.

However, if the size of the metadata store is greater than 0, then we need to recreate the metadata for those values. It will for sure be costly for these kinds of maps, but that cost is required for correctness. Either we pay the cost or make the necessary change to make the code not deserialize the HJV values by adding it to here https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/map/impl/record/Records.java#L226. This is much more simple to implement, but when you consider the indexed query performance of the maps that have BINARY/NATIVE in-memory format, the current approach will be much faster once the correct metadata is created. Check the binary and object in-memory format JSON query performance here, https://hazelcast.atlassian.net/wiki/spaces/IMDG/pages/907345951/JSON+Non-Indexed+Query+Optimisation there is approx. 5x difference, so I went with the current approach